internal/handlers/handler_verify_test.go

```diff
@@ -722,3 +735,311 @@ func TestSchemeIsWSS(t *testing.T) {
 	assert.True(t, isSchemeWSS(
 		GetURL("wss://mytest.example.com/abc/?query=abc")))
 }
+
+func TestShouldNotRefreshUserGroupsFromBackend(t *testing.T) {
+	mock := mocks.NewMockAutheliaCtx(t)
+	defer mock.Close()
+
+	// Setup pointer to john so we can adjust it during the test.
+	user := &authentication.UserDetails{
+		Username: "john",
+		Groups: []string{
+			"admin",
+			"users",
+		},
+		Emails: []string{
+			"john@example.com",
+		},
+	}
+
+	clock := mocks.TestingClock{}
+	clock.Set(time.Now())
+
+	userSession := mock.Ctx.GetSession()
+	userSession.Username = user.Username
+	userSession.AuthenticationLevel = authentication.TwoFactor
+	userSession.LastActivity = clock.Now().Unix()
+	userSession.Groups = user.Groups
+	userSession.Emails = user.Emails
+	userSession.KeepMeLoggedIn = true
+	err := mock.Ctx.SaveSession(userSession)
+	require.NoError(t, err)
+
+	mock.UserProviderMock.
+		EXPECT().
+		GetRefreshSettings().
+		Return(false, 0*time.Second)
+
+	mock.Ctx.Request.Header.Set("X-Original-URL", "https://two-factor.example.com")
+	VerifyGet(mock.Ctx)
+	assert.Equal(t, 200, mock.Ctx.Response.StatusCode())
+
+	// Request should get refresh settings and not get new user details.
+	mock.UserProviderMock.
+		EXPECT().
+		GetRefreshSettings().
+		Return(false, 0*time.Minute)
+
+	mock.Ctx.Request.Header.Set("X-Original-URL", "https://admin.example.com")
+	VerifyGet(mock.Ctx)
+	assert.Equal(t, 200, mock.Ctx.Response.StatusCode())
+
+	// Check Refresh TTL has not been updated.
+	userSession = mock.Ctx.GetSession()
+
+	// Check user groups are correct.
+	require.Len(t, userSession.Groups, len(user.Groups))
+	assert.Equal(t, utils.RFC3339Zero, userSession.RefreshTTL.Unix())
+	assert.Equal(t, "admin", userSession.Groups[0])
+	assert.Equal(t, "users", userSession.Groups[1])
+
+	mock.UserProviderMock.
+		EXPECT().
+		GetRefreshSettings().
+		Return(false, 0*time.Minute)
+
+	mock.Ctx.Request.Header.Set("X-Original-URL", "https://admin.example.com")
```

Comment:
Above all you need to emphasize there was no call to GetDetails(). You can use such construct to do that `EXPECT_CALL( mockObj, Foo(_) ).Times(0);`

Reply:
gomock actually does this by default if you properly implement it. However I added the check.